pennylane/gradients/hamiltonian_grad.py

```diff
@@ -26,14 +26,13 @@ def hamiltonian_grad(tape, idx):
     """
 
     op, m_pos, p_idx = tape.get_operation(idx)
-    new_tape = tape.copy(copy_operations=True)
 
     # get position in queue
     queue_position = m_pos - len(tape.operations)
-    new_tape._measurements[queue_position] = qml.expval(op.ops[p_idx])
+    new_measurements = list(tape.measurements)
+    new_measurements[queue_position] = qml.expval(op.ops[p_idx])
 
-    new_tape._par_info = {}
-    new_tape._update()
+    new_tape = qml.tape.QuantumScript(tape.operations, new_measurements, shots=tape.shots)
```

Comment:
Just to quiz myself, we don't care about setting the trainable parameters here because the trainable param index is already provided as an argument to `hamiltonian_grad`?

Reply:
The number of parameters if the script changes here, so `trainable_params` in not valid on the new tape.